### PR TITLE
Treat `Settings` object passed to `wandb.init()` similar to how `dict` is treated

### LINKED ISSUE #3505: [CLI]: _disable_stats doesn't work
### Describe the bug

`_disable_stats` doesn't work. `wandb.init(settings=wandb.Settings(_disable_stats=True))` It still sends stats to WANDB, which in turn leads to BSOD due to incompatibility with the old PYNVML dependency in the vendor folder in very specific scenarios.

```
wandb.init(settings=wandb.Settings(_disable_stats=True))
```

None

### Additional Files

_No response_

### Environment

WandB version:
0.12.14

OS:
Windows 10

Python version:
 3.9.10

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dmitryduev**
Fixes #3505

Description
-----------
Treat a `wandb.Settings` object passed to `wandb.init()` similar to how a passed `dict` is treated. Will now detect settings that differ from defaults and apply them using `Source.INIT`, as this is probably that the user would expect when doing `wandb.init(settings=wandb.Settings(...))`.

Also deprecated `config_include_keys` and `config_exclude_keys` `kwargs` in `wandb.init(...)` bc saw a todo in the code next to what I was touching and couldn't ignore it :)

Testing
-------
Hopefully fixed a couple tests + improved robustness in `tests/test_redir.py` that would flake out at times.

### CODE REVIEW

File: tests/wandb_run_test.py [hunk 1]
```diff
@@ -317,15 +317,21 @@ def test_use_artifact_offline(live_mock_server, test_settings):
     run.finish()
 
 
-def test_run_urls(test_settings):
+@pytest.mark.skip(
+    reason=(
+        "mock server enforces entity=mock_server_entity and project=test,"
+        "this should only work on a live server, or mock_server would need customization."
+    )
+)
```
Comment (dmitryduev) [inline tests/wandb_run_test.py:325]: This used to work bc previously, a Settings object passed to `wandb.init()` would "carry over" its setting sources. settings.update(...) by default uses Source.OVERRIDE meaning that no downstream changes would affect the changed settings (unless again overridden). 
We could customize `mock_server` so that it "recognizes" two (or more) entities, but it feels a bit outside the scope for this PR.

File: tests/wandb_settings_test.py [hunk 2]
```diff
@@ -995,3 +995,13 @@ def test_local_api_key_validation():
 
     # ensure that base_url is applied first without causing an error in api_key validation
     wandb.Settings()._apply_settings(s)
+
+
+def test_run_urls():
```
Comment (dmitryduev) [inline tests/wandb_settings_test.py:1000]: This is a "replacement" test for the one in `wandb_run_test`.

File: wandb/sdk/wandb_init.py [hunk 3]
```diff
@@ -138,10 +141,23 @@ def setup(self, kwargs) -> None:  # noqa: C901
         settings_param = kwargs.pop("settings", None)
         if settings_param is not None:
             if isinstance(settings_param, Settings):
-                # todo: check the logic here. this _only_ comes up in tests?
```
Comment (dmitryduev) [inline wandb/sdk/wandb_init.py:141]: see, I even left a todo back in the day lol :)

File: wandb/sdk/wandb_settings.py [hunk 4]
```diff
@@ -81,7 +82,7 @@ def _get_wandb_dir(root_dir: str) -> str:
     return os.path.expanduser(path)
 
 
-# fixme: should either return bool or error out. fix once confident.
```
Comment (dmitryduev) [inline wandb/sdk/wandb_settings.py:84]: fixme's break our flake checker, according to @raubitsj, so rm'ed 'em all.

File: tests/test_redir.py [hunk 5]
```diff
@@ -283,47 +284,51 @@ def test_offline_compression(test_settings, capfd, runner, console):
 
 @pytest.mark.parametrize("console", console_modes)
 @pytest.mark.parametrize("numpy", [True, False])
-@pytest.mark.timeout(120)
+@pytest.mark.timeout(300)
```
Comment (kptkin) [inline tests/test_redir.py:287]: why do you need a longer timeout?

Comment (dmitryduev) [reply tests/test_redir.py:287]: It failed for me locally once, so I ++'ed it

File: tests/test_redir.py [hunk 6]
```diff
@@ -236,16 +237,16 @@ def test_run_with_console_redirect(test_settings, capfd, console):
 
 
 @pytest.mark.parametrize("console", console_modes)
+@pytest.mark.timeout(300)
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply tests/test_redir.py]: oops, this one's not necessary, thanks! rm'ed

File: tests/wandb_settings_test.py (hunk 2 above)
Comment (kptkin) [reply tests/wandb_settings_test.py:1000]: do you mean `test_run_urls`?

File: wandb/sdk/wandb_init.py [hunk 7]
```diff
@@ -136,16 +137,8 @@ def setup(self, kwargs) -> None:  # noqa: C901
         # Start with settings from wandb library singleton
         settings: Settings = self._wl.settings.copy()
         settings_param = kwargs.pop("settings", None)
-        if settings_param is not None:
-            if isinstance(settings_param, Settings):
-                # todo: check the logic here. this _only_ comes up in tests?
-                # update settings with settings_param using whatever
-                # source each parameter has there
-                settings._apply_settings(settings_param, _logger=logger)
```
Comment (kptkin) [inline wandb/sdk/wandb_init.py:144]: what about this function `_apply_settings` it is not relevant here anymore?
(I see it is also used in `wandb_setup` and `wandb_login`)

File: wandb/sdk/wandb_init.py [hunk 8]
```diff
@@ -261,14 +258,8 @@ def setup(self, kwargs) -> None:  # noqa: C901
...
         # TODO(jhr): should this be moved? probably.
-        time_stamp: float = time.time()
-        settings.update(
-            {
-                "_start_time": time_stamp,
-                "_start_datetime": datetime.datetime.fromtimestamp(time_stamp),
-            },
-            source=Source.INIT,
-        )
+        # if settings._start_time is None:
```
Comment (kptkin) [inline wandb/sdk/wandb_init.py:261]: do you need this commented out code?

File: wandb/sdk/wandb_run.py [hunk 9]
```diff
@@ -408,7 +408,7 @@ def _init(
         self._step = 0
         self._torch_history: Optional["wandb.wandb_torch.TorchHistory"] = None
 
-        self._start_time = time.time()
+        self._start_time = self._settings._start_time or time.time()
```
Comment (kptkin) [inline wandb/sdk/wandb_run.py:411]: I think this one is on purpose (not sure) cc: @raubitsj

File: wandb/sdk/wandb_settings.py [hunk 10]
```diff
@@ -636,9 +637,11 @@ def _default_props(self) -> Dict[str, Dict[str, Any]]:
             system_sample_seconds={"value": 2},
             timespec={
                 "hook": (
-                    lambda _: datetime.strftime(self._start_datetime, "%Y%m%d_%H%M%S")
-                    if self._start_time and self._start_datetime
-                    else None
+                    lambda _: (
+                        datetime.strftime(self._start_datetime, "%Y%m%d_%H%M%S")
+                        if self._start_datetime
```
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:642]: yeah i was wondering about this condition :)

File: wandb/sdk/wandb_settings.py [hunk 11]
```diff
@@ -917,11 +920,16 @@ def _run_url(self) -> str:
         query = self._get_url_query_string()
         return f"{project_url}/runs/{quote(self.run_id)}{query}"
 
-    def _start_run(self) -> None:
+    def _start_run(self, source: int = Source.BASE) -> None:
```
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:923]: nit: name seems weird given that we only time stamp this

File: wandb/sdk/wandb_init.py [hunk 12]
```diff
@@ -180,13 +173,17 @@ def setup(self, kwargs) -> None:  # noqa: C901
...
+        deprecated_kwargs = {
+            "config_include_keys": (
+                "Use `config=wandb.helper.parse_config(config_object, include=('key',))` instead."
+            ),
+            "config_exclude_keys": (
+                "Use `config=wandb.helper.parse_config(config_object, exclude=('key',))` instead."
+            ),
+        }
+        for deprecated_kwarg, msg in deprecated_kwargs.items():
+            if locals().get(deprecated_kwarg):
```
Comment (kptkin) [inline wandb/sdk/wandb_init.py:185]: I think there is a problem using `local()`. See this PR where we wrap it in `dict`:
https://github.com/wandb/client/pull/1390

Comment (kptkin) [inline wandb/sdk/wandb_init.py:185]: something like this?
```suggestion
        local_args = dict(locals())
        for deprecated_kwarg, msg in deprecated_kwargs.items():
            if deprecated_kwarg in local_args:
```

File: tests/test_redir.py (hunk 5 above)
Comment (kptkin) [reply tests/test_redir.py:286]: ok if you think it makes sense :)

File: tests/utils/utils.py [hunk 13]
```diff
@@ -165,3 +167,26 @@ def magic(path, *args, **kwargs):
     env["KUBERNETES_PORT_443_TCP_PORT"] = "123"
     env["HOSTNAME"] = "test"
     return env
+
+
+def retry(retries: int = 3, delay: int = 1) -> Callable:
```
Comment (kptkin) [inline tests/utils/utils.py:172]: is this function used anywhere?

Comment (dmitryduev) [reply tests/utils/utils.py:172]: No (it was used in something that turned out to be unnecessary). do you think I should remove it?

File: wandb/sdk/wandb_settings.py (hunk 11 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_settings.py:923]: Agreed. I just kept the way it was called previously. How about `_set_start_times`?

File: wandb/sdk/wandb_run.py (hunk 9 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_run.py:411]: ah, right. I think I remember someone even complaining about it. @raubitsj should I revert it?

File: wandb/sdk/wandb_init.py (hunk 7 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_init.py:144]: Right, not used here anymore as the logic changed, but still used in other places. 
(`_apply_settings` applies settings to the target object using the Sources in the "source" object, so only the settings with greater-than-or-equal-to priorities will "pass"; in this case we want to explicitly apply `Source.INIT` as per user expectations)

Comment (kptkin) [reply wandb/sdk/wandb_init.py:144]: Got it! Thanks for the explanation

File: wandb/sdk/wandb_init.py (hunk 12 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_init.py]: oh, totally forgot about that, thank you!!
rewrote this a bit, hopefully it's more clear now. (i was checking if the local `config_<in|ex>clude_keys` variables were not None). Wrote it this way in case we want to deprecate more kwargs in the future.

File: wandb/sdk/wandb_settings.py (hunk 11 above)
Comment (kptkin) [reply wandb/sdk/wandb_settings.py]: yeah makes more sense...

File: tests/utils/utils.py (hunk 13 above)
Comment (kptkin) [reply tests/utils/utils.py:172]: hmmm... whatever you want... :)

Comment (dmitryduev) [reply tests/utils/utils.py:172]: I thought it might be useful one day in some tests, especially where we're relying on naping

Comment (dmitryduev) [reply tests/utils/utils.py:172]: we have the retry instruments in `lib`, but they are too heavy/complex for cases like that

**Review by kptkin — APPROVED** [review]
Looks good!

File: tests/utils/utils.py (hunk 13 above)
Comment (dmitryduev) [reply tests/utils/utils.py:172]: actually, let me yank it since it makes codecov go down :)

File: tests/conftest.py [hunk 14]
```diff
@@ -225,9 +225,11 @@ def test_settings(test_dir, mocker, live_mock_server):
...
     mkdir_exists_ok(wandb_dir)
     # root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+    time_stamp: float = time.time()
+    datetime_now: datetime.datetime = datetime.datetime.fromtimestamp(time_stamp)
     settings = wandb.Settings(
-        _start_datetime=datetime.datetime.now(),
-        _start_time=time.time(),
+        _start_datetime=datetime_now,
+        _start_time=time_stamp,
         api_key=DUMMY_API_KEY,
```
Comment (kptkin) [inline tests/conftest.py:233]: curious: why we can't use `settings._set_time_stamps()` here?

File: wandb/sdk/wandb_settings.py (hunk 11 above)
Comment (dmitryduev) [reply wandb/sdk/wandb_settings.py]: I'd keep a private method, but would be happy to move the call in `wandb.init()` to `_apply_init`. @raubitsj there's a todo in that place from you saying "# TODO(jhr): should this be moved? probably.", so I'd defer it to you to make the final decision thanks :)
Also, it's used in `standalone_tests/grpc_client.py`.

File: tests/conftest.py (hunk 14 above)
Comment (dmitryduev) [reply tests/conftest.py:233]: sure, good call.

File: wandb/sdk/wandb_run.py (hunk 9 above)
Comment (raubitsj) [reply wandb/sdk/wandb_run.py:411]: I think andrew wanted it for importing a run but it doesnt feel like the best way to support this... lets revert this for now unless you tested all the downstream parts of this... if you set start time to 2 days ago and run a test for 15 minutes.. does the total time get calculated right?  my guess without looking at the code fully is that the runtime might be 2days and 15 minutes.

Comment (dmitryduev) [reply wandb/sdk/wandb_run.py:411]: ah, yea, it was Andrew, right! Got it, let me revert for now, it's not related to this PR (but will add a todo).
